File: algorithms/0503.next_greater_element_ii/next_greater_element_ii.rs
impl Solution {
    pub fn next_greater_elements(nums: Vec<i32>) -> Vec<i32> {
        let mut ans = vec![-1; nums.len()];
        let mut greater_stack = Vec::<usize>::new();
        for i in 0..(nums.len() * 2) {
            while !greater_stack.is_empty()
                && nums[*greater_stack.last().unwrap()] < nums[i % nums.len()]
            {
                ans[greater_stack.pop().unwrap()] = nums[i % nums.len()];
            }
            greater_stack.push(i % nums.len());
        }

        ans
    }
}

File: algorithms/0503.next_greater_element_ii/next_greater_element_ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        const int len = nums.size();
        vector<int> greater_nums(len, -1);
        stack<int> mono_stack;
        stack<int> index_stack;
        for (int i = 0; i < len; ++i) {
            nums.push_back(nums[i]);
        }
        for (int i = 0; i < nums.size(); ++i) {
            const int &n = nums[i];
            while (!mono_stack.empty() && n > mono_stack.top()) {
                mono_stack.pop();
                const int prev_index = index_stack.top() % len; index_stack.pop();
                greater_nums[prev_index] = n;
            }
            mono_stack.push(n); index_stack.push(i);
        }
        return greater_nums;
    }
};